src/hooks/use-query.tsx

```diff
@@ -1,28 +1,104 @@
 import {
-  useQuery as useReactQuery,
-  QueryKey,
-  UseQueryOptions
+  type DefinedInitialDataOptions,
+  type DefinedUseQueryResult,
+  type QueryKey,
+  type UndefinedInitialDataOptions,
+  type UseQueryOptions,
+  type UseQueryResult,
+  useQuery as useReactQuery
 } from '@tanstack/react-query'
-import { ErrorResponse } from '~/types'
 
-interface UseQueryProps<TData, TError, TSelect> {
-  queryKey: QueryKey
-  queryFn: () => Promise<TData>
-  options?: UseQueryOptions<TData, TError, TSelect>
-}
+import { type ResponseError } from '~/exceptions'
+
+type RequiredQueryOptions = 'queryFn' | 'queryKey'
+type OptionalQueryOptions = 'enabled' | 'initialData' | 'select' | 'staleTime'
+type ReturningQueryParams =
+  | 'data'
+  | 'error'
+  | 'isError'
+  | 'isFetching'
+  | 'isLoading'
+  | 'refetch'
+
+type RequiredQueryProperties<
+  TQueryFnData,
+  TError = ResponseError,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey
+> = Required<
+  Pick<
+    UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    RequiredQueryOptions
+  >
+>
+
+type Properties<
+  TQueryFnData,
+  TError = ResponseError,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey
+> = {
+  options?: Pick<
+    UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
+    OptionalQueryOptions
+  >
+} & RequiredQueryProperties<TQueryFnData, TError, TData, TQueryKey>
+
+function useQuery<
+  TQueryFnData,
+  TError = ResponseError,
+  TData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey
+>(
+  properties: {
+    options: Pick<
+      DefinedInitialDataOptions<TQueryFnData, TError, TData, TQueryKey>,
+      OptionalQueryOptions
+    >
+  } & RequiredQueryProperties<TQueryFnData, TError, TData, TQueryKey>
+): Pick<DefinedUseQueryResult<TData, TError>, ReturningQueryParams>
```

Comment:
maybe you should move the long type definitions into separate types, so that the useQuery definition won't be as accumulated